File: src/main/java/gargoyle/ct/resource/impl/CTConfigResource.java
package gargoyle.ct.resource.impl;

import gargoyle.ct.config.CTConfig;
import gargoyle.ct.ex.CTException;
import gargoyle.ct.resource.Resource;
import gargoyle.ct.resource.internal.LocalResource;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Objects;

/**
 * {@link CTConfig} as {@link Resource}
 */
public final class CTConfigResource extends LocalResource {
    private static final String SLASH = "/";

    private CTConfigResource(@NotNull URL url) {
        super(url);
    }

    @Nullable
    public static CTConfigResource findLocalConfig(@NotNull String name, boolean existing) {
        LocalResource local = LocalResource.findLocal(name);
        try {
            return (local == null) ? null : ((existing && !local.exists()) ? null : new CTConfigResource(Objects.requireNonNull(local.toURL())));
        } catch (IOException ex) {
            return null;
        }
    }

    public static CTConfigResource forURL(@NotNull URL url) {
        return new CTConfigResource(url);
    }

    public static CTConfigResource forURL(URL root, String file) {
        try {
            String rootString = root.toExternalForm();
            return new CTConfigResource(new URL((rootString.endsWith(SLASH) ? rootString : rootString + SLASH) + file));
        } catch (MalformedURLException ex) {
            throw new CTException(file, ex);
        }
    }
}

File: app/src/main/kotlin/gargoyle/ct/resource/impl/CTConfigResource.kt
package gargoyle.ct.resource.impl

import gargoyle.ct.config.CTConfig
import gargoyle.ct.ex.CTException
import gargoyle.ct.util.resource.internal.LocalResource
import java.io.IOException
import java.net.MalformedURLException
import java.net.URL

/**
 * [CTConfig] as [Resource]
 */
class CTConfigResource private constructor(url: URL) : LocalResource(url) {
    companion object {

        fun findLocalConfig(name: String, existing: Boolean): CTConfigResource? =
            findLocal(name).let {
                try {
                    if (existing && !it.exists()) null else CTConfigResource(requireNotNull(it.toURL()))
                } catch (ex: IOException) {
                    null
                }
            }

        fun forURL(url: URL): CTConfigResource = CTConfigResource(url)

        fun forURL(root: URL, file: String): CTConfigResource =
            try {
                CTConfigResource(URL("${root.toExternalForm().trimEnd('/')}/$file"))
            } catch (ex: MalformedURLException) {
                throw CTException(file, ex)
            }
    }
}
